pkg/cimfs/cim_test.go

```diff
@@ -0,0 +1,146 @@
+//go:build windows
+// +build windows
+
+package cimfs
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/osversion"
+	"golang.org/x/sys/windows"
+)
+
+// A simple tuple type used to hold information about a file/directory that is created
+// during a test.
+type tuple struct {
+	filepath     string
+	fileContents []byte
+	isDir        bool
+}
+
+// A utility function to create a file/directory and write data to it in the given cim
+func createCimFileUtil(c *CimFsWriter, fileTuple tuple) error {
+	// create files inside the cim
+	fileInfo := &winio.FileBasicInfo{
+		CreationTime:   windows.NsecToFiletime(time.Now().UnixNano()),
+		LastAccessTime: windows.NsecToFiletime(time.Now().UnixNano()),
+		LastWriteTime:  windows.NsecToFiletime(time.Now().UnixNano()),
+		ChangeTime:     windows.NsecToFiletime(time.Now().UnixNano()),
+		FileAttributes: 0,
+	}
+	if fileTuple.isDir {
+		fileInfo.FileAttributes = windows.FILE_ATTRIBUTE_DIRECTORY
+	}
+
+	if err := c.AddFile(filepath.FromSlash(fileTuple.filepath), fileInfo, int64(len(fileTuple.fileContents)), []byte{}, []byte{}, []byte{}); err != nil {
+		return err
+	}
+
+	if !fileTuple.isDir {
+		wc, err := c.Write(fileTuple.fileContents)
+		if err != nil || wc != len(fileTuple.fileContents) {
+			if err == nil {
+				return fmt.Errorf("unable to finish writing to file %s", fileTuple.filepath)
+			} else {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// This test creates a cim, writes some files to it and then reads those files back.
+// The cim created by this test has only 3 files in the following tree
+// /
+// |- foobar.txt
+// |- foo
+// |--- bar.txt
+func TestCimReadWrite(t *testing.T) {
+	if osversion.Get().Build < MinimumCimFSBuild {
+		t.Skipf("Requires build %d+", MinimumCimFSBuild)
+	}
+
+	testContents := []tuple{
+		{"foobar.txt", []byte("foobar test data"), false},
+		{"foo", []byte(""), true},
+		{"foo\\bar.txt", []byte("bar test data"), false},
+	}
+
+	tempDir := t.TempDir()
+
+	cimName := "test.cim"
+	cimPath := filepath.Join(tempDir, cimName)
+	c, err := Create(tempDir, "", cimName)
+	if err != nil {
+		t.Fatalf("failed while creating a cim: %s", err)
+	}
+	defer func() {
+		// destroy cim sometimes fails if tried immediately after accessing & unmount the cim so give
+		// some time and try again.  A fix for this is in progress and we should be able to remove
+		// this sleep after it gets merged.
```

Comment:
What is the timeline for this fix?

Reply:
Actually, this is by design and can't be easily changed. I have removed the comment and kept the sleep.